src/torchcodec/_core/Encoder.cpp

```diff
@@ -0,0 +1,214 @@
+#include "src/torchcodec/_core/Encoder.h"
+#include "torch/types.h"
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+}
+
+namespace facebook::torchcodec {
+
+Encoder::~Encoder() {}
+
+// TODO-ENCODING: disable ffmpeg logs by default
+
+Encoder::Encoder(
+    const torch::Tensor wf,
+    int sampleRate,
+    std::string_view fileName)
+    : wf_(wf), sampleRate_(sampleRate) {
+  TORCH_CHECK(
+      wf_.dtype() == torch::kFloat32,
+      "waveform must have float32 dtype, got ",
+      wf_.dtype());
+  TORCH_CHECK(
+      wf_.dim() == 2, "waveform must have 2 dimensions, got ", wf_.dim());
+  AVFormatContext* avFormatContext = nullptr;
+  avformat_alloc_output_context2(
+      &avFormatContext, nullptr, nullptr, fileName.data());
+  TORCH_CHECK(
+      avFormatContext != nullptr,
+      "Couldn't allocate AVFormatContext. ",
+      "Check the desired extension?");
```

Comment:
We should check the return value of `avformat_alloc_output_context2()`; that should enable us to return a more useful error message. See https://ffmpeg.org/doxygen/3.3/avformat_8h.html#a6ddf3d982feb45fa5081420ee911f5d5.